### PR TITLE
Test case for cluster templates

### PR DESCRIPTION
**Pull request by soumyalj**
@sangeethah : Please review. 
https://github.com/rancher/rancher/issues/23328

### CODE REVIEW

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+from rancher import ApiError, Client
+
+DO_ACCESSKEY = os.environ.get('DO_ACCESSKEY', "None")
+RANCHER_S3_BUCKETNAME = os.environ.get('RANCHER_S3_BUCKETNAME', "None")
+RANCHER_S3_ENDPOINT = os.environ.get('RANCHER_S3_ENDPOINT', "None")
+AWS_ACCESS_KEY_ID = os.environ.get('AWS_ACCESS_KEY_ID', "None")
+AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY', "None")
+RANCHER_DEFAULT_K8S_VERSION = os.environ.get('RANCHER_DEFAULT_K8S_VERSION',
+                                             "v1.14.8-rancher1-1")
+
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: v3/settings/k8s-versions-current should have the list of supported K8s versions . This variable should take values from here.

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        "enableClusterAlerting": "false",
+        "enableClusterMonitoring": "false",
+        "enableNetworkPolicy": "false",
+        "type": "clusterSpecBase",
+        "localClusterAuthEndpoint": {
+            "enabled": "true",
+            "type": "localClusterAuthEndpoint"
+        },
+        "rancherKubernetesEngineConfig": rke_config
+    }
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: use get_cluster_config() method instead

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        standard_newuser_client, name=cluster_name,
+        clusterTemplateRevisionId=cluster_template_revision1.id)
+
+    cluster_cleanup(standard_newuser_client, cluster)
+
+
+def cluster_template_create_edit(testclient):
+
+    cluster_config1 = get_cluster_config(RANCHER_DEFAULT_K8S_VERSION)
+    cluster_config2 = get_cluster_config("v1.15.4-rancher1-2")
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: This value needs to be got from `k8s-versions-current`

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+    cluster_name = random_test_name("test-auto")
+    cluster = create_node_cluster(
+        client, name=cluster_name,
+        clusterTemplateRevisionId=cluster_template_revision1.id)
+
+    cluster_template_revision2 = client.create_cluster_template_revision(
+        name=random_test_name("revision2"),
+        clusterConfig=cluster_config2,
+        clusterTemplateId=cluster_template.id)
+
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: Assert k8s versions of the cluster using kubectl command

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+    cluster = client.update(
+            cluster,
+            name=cluster_name,
+            clusterTemplateRevisionId=cluster_template_revision2.id)
+
+    cluster = validate_cluster(client,
+                               cluster,
+                               intermediate_state="updating")
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: Assert k8s versions of the cluster using kubectl command

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+
+def test_cluster_template_create_edit_stduser():
+
+    standard_user_client = Client(
+                         url=CATTLE_API_URL,
+                         token=user_token["user_standard"]["token"],
+                         verify=False)
+    cluster_template_create_edit(standard_user_client)
+
+
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: Add test description that outlines the workflow

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    cluster_template_revision = \
+        standard_newuser_client.create_cluster_template_revision(
+            name=revision_name,
+            clusterConfig=cluster_config1,
+            clusterTemplateId=cluster_template.id)
+    cluster_name = "test-auto"
+    cluster = create_node_cluster(
+        standard_newuser_client, name=cluster_name,
+        clusterTemplateRevisionId=cluster_template_revision.id)
+
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: Create another CT with 1 CTR and make sure that "standard_newuser_client" is not able to create new revisions for this CT or deploy a cluster with existing CTR.

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+    assert e.value.error.status == 403
+    assert e.value.error.code == 'PermissionDenied'
+
+    # Verify that the read-only user can create a cluster with the existing
+    # template revision
+    cluster_name = "test-auto"
+    cluster = create_node_cluster(
+        standard_newuser_client, name=cluster_name,
+        clusterTemplateRevisionId=cluster_template_revision1.id)
+
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_cluster_templates.py]: As admin Create another CT with 1 CTR and make sure that "standard_newuser_client" is not able deploy a cluster with existing CTR.

File: tests/validation/tests/v3_api/test_cluster_templates.py (hunk 1 above)
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: Done

File: tests/validation/tests/v3_api/test_cluster_templates.py (hunk 2 above)
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: Done

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+        "type": "string",
+        "default": "s3.amazonaws.com"
+
+        },
+        {
+        "variable": "rancherKubernetesEngineConfig.services.etcd.backupConfig."
+                    "s3BackupConfig.accessKey",
+        "required": "true",
+        "type": "string",
+        "default": "AKBBBBBB"
```
[the remark this answers is not in the document]
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: UI fills in the default field with the value entered for the respective field. We can use any random value/use empty for the S3 config parameters which will be corrected as answers .

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+        },
+        {
+        "variable": "rancherKubernetesEngineConfig.services.etcd.backupConfig."
+                    "s3BackupConfig.secretKey",
+        "required": "true",
+        "type": "string",
+        "default": "BNNNNNNNNN"
```
[the remark this answers is not in the document]
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: UI fills in the default field with the value entered for the respective field. We can use any random value/use empty for the S3 config parameters which will be corrected as answers .

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+                "s3BackupConfig.accessKey": AWS_ACCESS_KEY_ID,
+            "rancherKubernetesEngineConfig.services.etcd.backupConfig."
+                "s3BackupConfig.secretKey": AWS_SECRET_ACCESS_KEY
+        }
+    }
+
+    # client = get_admin_client()
+    standard_user_client = Client(url=CATTLE_API_URL,
+                                  token=user_token["user_standard"]["token"],
+                                  verify=False)
```
[the remark this answers is not in the document]
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: Done

File: tests/validation/tests/v3_api/test_cluster_templates.py [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+    assert cluster.appliedSpec.rancherKubernetesEngineConfig.services.etcd.\
+        backupConfig.s3BackupConfig.bucketName == RANCHER_S3_BUCKETNAME
+    assert cluster.appliedSpec.rancherKubernetesEngineConfig.services.\
+        etcd.backupConfig.s3BackupConfig.endpoint == RANCHER_S3_ENDPOINT
+    assert cluster.appliedSpec.rancherKubernetesEngineConfig.services.etcd.\
+        backupConfig.s3BackupConfig.accessKey == AWS_ACCESS_KEY_ID
+    assert cluster.appliedSpec.rancherKubernetesEngineConfig.services.etcd.\
+        backupConfig.s3BackupConfig.type == "/v3/schemas/s3BackupConfig"
+    assert cluster.appliedSpec.rancherKubernetesEngineConfig.network.plugin ==\
+        "flannel"
```
[the remark this answers is not in the document]
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: docker directory and localClusterAuthEndpoint parameters are asserted .

File: tests/validation/tests/v3_api/test_cluster_templates.py (hunk 8 above)
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: Done

File: tests/validation/tests/v3_api/test_cluster_templates.py (hunk 4 above)
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: check_cluster_version(cluster,version) method in common.py

File: tests/validation/tests/v3_api/test_cluster_templates.py (hunk 5 above)
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: check_cluster_version(cluster,version) method in common.py

File: tests/validation/tests/v3_api/test_cluster_templates.py (hunk 6 above)
Comment (soumyalj) [reply tests/validation/tests/v3_api/test_cluster_templates.py]: Done

**Comment by sangeethah** [comment]
LGTM

**Pull request merged**